Drop SharedMemory::data() in favor of span()

cdumez opened the pull request:
#### 78d8c59ed298b1c74189daf439a9104dbb7b2e2c
<pre>
Drop SharedMemory::data() in favor of span()
<a href="https://bugs.webkit.org/show_bug.cgi?id=274943">https://bugs.webkit.org/show_bug.cgi?id=274943</a>

Reviewed by Darin Adler.

* Source/WebCore/PlatformMac.cmake:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/ShareableResource.cpp:
(WebCore::ShareableResource::data const):
* Source/WebCore/platform/SharedMemory.cpp:
(WebCore::SharedMemory::copyBuffer):
(WebCore::SharedMemory::createSharedBuffer const):
* Source/WebCore/platform/SharedMemory.h:
(WebCore::SharedMemory::size const):
(WebCore::SharedMemory::data const): Deleted.
* Source/WebCore/platform/cocoa/SharedMemoryCocoa.mm: Renamed from Source/WebCore/platform/cocoa/SharedMemoryCocoa.cpp.
(WebCore::toVMMemoryLedger):
(WebCore::SharedMemoryHandle::takeOwnershipOfMemory const):
(WebCore::SharedMemoryHandle::setOwnershipOfMemory const):
(WebCore::toPointer):
(WebCore::toVMAddress):
(WebCore::SharedMemory::allocate):
(WebCore::machProtection):
(WebCore::makeMemoryEntry):
(WebCore::SharedMemory::wrapMap):
(WebCore::SharedMemory::map):
(WebCore::SharedMemory::~SharedMemory):
(WebCore::SharedMemory::createHandle):
(WebCore::SharedMemory::createSendRight const):
(WebCore::SharedMemory::toNSData const):
* Source/WebCore/platform/graphics/ShareableBitmap.cpp:
(WebCore::ShareableBitmap::data const):
* Source/WebCore/platform/graphics/avfoundation/objc/CDMSessionAVContentKeySession.mm:
* Source/WebCore/platform/graphics/cg/ShareableBitmapCG.mm:
(WebCore::ShareableBitmap::createFromImagePixels):
* Source/WebCore/platform/graphics/mac/controls/SwitchMacUtilities.h:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.cpp:
(WebKit::RemoteGraphicsContextGL::readPixelsSharedMemory):
* Source/WebKit/GPUProcess/graphics/RemoteImageBuffer.cpp:
(WebKit::RemoteImageBuffer::getPixelBuffer):
* Source/WebKit/GPUProcess/media/RemoteAudioDestinationManager.cpp:
* Source/WebKit/Platform/IPC/StreamConnectionBuffer.cpp:
(IPC::StreamConnectionBuffer::headerForTesting):
* Source/WebKit/Platform/IPC/StreamConnectionBuffer.h:
(IPC::StreamConnectionBuffer::data const):
(IPC::StreamConnectionBuffer::header const):
* Source/WebKit/Shared/Cocoa/SharedCARingBuffer.cpp:
(WebKit::ProducerSharedCARingBuffer::allocate):
* Source/WebKit/Shared/Cocoa/SharedCARingBuffer.h:
* Source/WebKit/Shared/IPCStreamTester.cpp:
(WebKit::IPCStreamTester::syncMessageReturningSharedMemory1):
* Source/WebKit/Shared/SharedStringHashStore.cpp:
(WebKit::SharedStringHashStore::resizeTable):
* Source/WebKit/Shared/SharedStringHashTable.cpp:
(WebKit::SharedStringHashTable::clear):
* Source/WebKit/Shared/SharedStringHashTableReadOnly.cpp:
(WebKit::SharedStringHashTableReadOnly::setSharedMemory):
* Source/WebKit/Shared/WebHitTestResultData.cpp:
(WebKit::WebHitTestResultData::getImageSharedMemoryHandle const):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(convertAndAddHighlight):
* Source/WebKit/UIProcess/mac/WebContextMenuProxyMac.mm:
(WebKit::WebContextMenuProxyMac::getShareMenuItem):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::getPixelBufferForImageBuffer):
* Source/WebKit/WebProcess/GPU/media/RemoteAudioDestinationProxy.cpp:
(WebKit::RemoteAudioDestinationProxy::totalFrameCount const):
* Source/WebKit/WebProcess/GPU/webrtc/SharedVideoFrame.cpp:
(WebKit::SharedVideoFrameWriter::writeBuffer):
* Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
(WebKit::IPCTestingAPI::JSSharedMemory::writeBytes):

Canonical link: <a href="https://commits.webkit.org/279598@main">https://commits.webkit.org/279598@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3fd439500acdeac2396230bfed58b927bf508168

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/53892 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/33262 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/6424 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/57172 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/4617 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/56196 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/40767 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/4492 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/43645 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/3043 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/55990 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/31483 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/46620 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/24781 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/28314 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/3940 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/2770 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/49976 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/4143 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/58765 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/29066 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/4218 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/51059 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/30258 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/46761 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/50398 "Found 1 new API test failure: /WebKitGTK/TestWebViewEditor:/webkit/WebKitWebView/select-all/non-editable (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/31204 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7983 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/30035 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/platform/cocoa/SharedMemoryCocoa.mm:
```diff
@@ -228,4 +228,10 @@ WTF::MachSendRight SharedMemory::createSendRight(Protection protection) const
+RetainPtr<NSData> SharedMemory::toNSData() const
+{
+    auto span = this->span();
+    return [NSData dataWithBytes:span.data() length:span.size()];
```
Until we switch to ARC, this does an `autorelease`, which may extend the lifetime of the `NSData` object even past the lifetime of the `SharedMemory` object in some cases. So I’d consider `initWithBytes:length:` instead.

darinadler commented on Source/WebCore/platform/graphics/mac/controls/SwitchMacUtilities.h:
```diff
@@ -26,7 +26,20 @@
+
+struct ControlStyle;
+
+template<typename T> class RectEdges;
```
No need for the `T` here.

darinadler commented on Source/WebCore/platform/cocoa/SharedMemoryCocoa.mm:
```diff
@@ -228,4 +228,10 @@ WTF::MachSendRight SharedMemory::createSendRight(Protection protection) const
+RetainPtr<NSData> SharedMemory::toNSData() const
+{
+    auto span = this->span();
+    return adoptNS([NSData initWithBytes:span.data() length:span.size()]);
```
Oops!
```
return adoptNS([[NSData alloc] initWithBytes:span.data() length:span.size()]);
```

cdumez committed:
Drop SharedMemory::data() in favor of span()
https://bugs.webkit.org/show_bug.cgi?id=274943

Reviewed by Darin Adler.

* Source/WebCore/PlatformMac.cmake:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/ShareableResource.cpp:
(WebCore::ShareableResource::data const):
* Source/WebCore/platform/SharedMemory.cpp:
(WebCore::SharedMemory::copyBuffer):
(WebCore::SharedMemory::createSharedBuffer const):
* Source/WebCore/platform/SharedMemory.h:
(WebCore::SharedMemory::size const):
(WebCore::SharedMemory::data const): Deleted.
* Source/WebCore/platform/cocoa/SharedMemoryCocoa.mm: Renamed from Source/WebCore/platform/cocoa/SharedMemoryCocoa.cpp.
(WebCore::toVMMemoryLedger):
(WebCore::SharedMemoryHandle::takeOwnershipOfMemory const):
(WebCore::SharedMemoryHandle::setOwnershipOfMemory const):
(WebCore::toPointer):
(WebCore::toVMAddress):
(WebCore::SharedMemory::allocate):
(WebCore::machProtection):
(WebCore::makeMemoryEntry):
(WebCore::SharedMemory::wrapMap):
(WebCore::SharedMemory::map):
(WebCore::SharedMemory::~SharedMemory):
(WebCore::SharedMemory::createHandle):
(WebCore::SharedMemory::createSendRight const):
(WebCore::SharedMemory::toNSData const):
* Source/WebCore/platform/graphics/ShareableBitmap.cpp:
(WebCore::ShareableBitmap::data const):
* Source/WebCore/platform/graphics/avfoundation/objc/CDMSessionAVContentKeySession.mm:
* Source/WebCore/platform/graphics/cg/ShareableBitmapCG.mm:
(WebCore::ShareableBitmap::createFromImagePixels):
* Source/WebCore/platform/graphics/mac/controls/SwitchMacUtilities.h:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.cpp:
(WebKit::RemoteGraphicsContextGL::readPixelsSharedMemory):
* Source/WebKit/GPUProcess/graphics/RemoteImageBuffer.cpp:
(WebKit::RemoteImageBuffer::getPixelBuffer):
* Source/WebKit/GPUProcess/media/RemoteAudioDestinationManager.cpp:
* Source/WebKit/Platform/IPC/StreamConnectionBuffer.cpp:
(IPC::StreamConnectionBuffer::headerForTesting):
* Source/WebKit/Platform/IPC/StreamConnectionBuffer.h:
(IPC::StreamConnectionBuffer::data const):
(IPC::StreamConnectionBuffer::header const):
* Source/WebKit/Shared/Cocoa/SharedCARingBuffer.cpp:
(WebKit::ProducerSharedCARingBuffer::allocate):
* Source/WebKit/Shared/Cocoa/SharedCARingBuffer.h:
* Source/WebKit/Shared/IPCStreamTester.cpp:
(WebKit::IPCStreamTester::syncMessageReturningSharedMemory1):
* Source/WebKit/Shared/SharedStringHashStore.cpp:
(WebKit::SharedStringHashStore::resizeTable):
* Source/WebKit/Shared/SharedStringHashTable.cpp:
(WebKit::SharedStringHashTable::clear):
* Source/WebKit/Shared/SharedStringHashTableReadOnly.cpp:
(WebKit::SharedStringHashTableReadOnly::setSharedMemory):
* Source/WebKit/Shared/WebHitTestResultData.cpp:
(WebKit::WebHitTestResultData::getImageSharedMemoryHandle const):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(convertAndAddHighlight):
* Source/WebKit/UIProcess/mac/WebContextMenuProxyMac.mm:
(WebKit::WebContextMenuProxyMac::getShareMenuItem):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::getPixelBufferForImageBuffer):
* Source/WebKit/WebProcess/GPU/media/RemoteAudioDestinationProxy.cpp:
(WebKit::RemoteAudioDestinationProxy::totalFrameCount const):
* Source/WebKit/WebProcess/GPU/webrtc/SharedVideoFrame.cpp:
(WebKit::SharedVideoFrameWriter::writeBuffer):
* Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
(WebKit::IPCTestingAPI::JSSharedMemory::writeBytes):

Canonical link: https://commits.webkit.org/279598@main
